feat: #400 import mapping

birtona opened the pull request:
still WIP

MarcusRiemer commented on app/models/import_mapping.rb:
```diff
@@ -5,15 +5,37 @@ class ImportMapping < ApplicationRecord
   validates :name, uniqueness: true
 
+  def parse(value, parser_names)
+    parsers = parser_names&.map { |parser_name| @@parsers[parser_name] } || []
+    parsers.each do |parser|
+      value = parser.call(value) if parser
```
Lieber expliziten Fehler werfen, wenn der Parser nicht existiert.

birtona replied:
ist ergänzt: https://github.com/a-thousand-channels/ORTE-backend/pull/406/commits/8d1a73945b4160b57baa80c6343ec138a6a07c2d

MarcusRiemer commented on app/views/import_mappings/new.html.haml:
```diff
@@ -0,0 +1,53 @@
+          .alert.alert-danger
+            %p= flash[:alert]
+
+    - if @missing_fields && !@missing_fields.empty?
```
`@missing_fields&.present?`

birtona replied:
changed

MarcusRiemer commented on app/views/import_mappings/new.html.haml:
```diff
@@ -0,0 +1,53 @@
+        const parser = row.querySelector('.parser-input').value;
+        mappings.push({ csv_column_name: header, model_property: modelProperty, parser: parser });
+      });
+      document.getElementById('import_mapping_mapping').value = JSON.stringify(mappings);
```
Alternativ: In Rails ein passendes Model für das Formular bauen. Bin aber nicht sicher, ob ich das schöner finde.

birtona replied:
wie besprochen aktuell nicht nötig

MarcusRiemer commented on app/controllers/layers_controller.rb:
```diff
@@ -25,15 +25,26 @@ def import; end
+      redirect_to new_import_mapping_path(missing_fields: e.missing_fields, headers: @headers)
     end
     @valid_rows = importer.valid_rows
     session[:importing_rows] = @valid_rows
```
Standardmäßig lebt die Session im Cookie mit 4 KB Größenbeschränkung.

birtona replied:
Hat sich erledigt weil wir jetzt nichts mehr in der Session speichern

MarcusRiemer commented on app/models/import_mapping.rb:
```diff
@@ -0,0 +1,65 @@
+    end.map do |column_name|
+      {
+        'csv_column_name' => column_name,
+        'model_property' => Place.column_names.find { |col| col.casecmp(column_name).zero? },
```
Nitpick: Ruby hat seit 2.4 `casecmp?` das direkt einen booleschen wert liefert.

birtona replied:
geändert

MarcusRiemer commented on app/models/import_mapping.rb:
```diff
@@ -0,0 +1,65 @@
+    parsers.each do |parser|
+      value = parser.call(value) if parser
+    end
+    value
```
Expliziter Fehler für fehlende Parser?

birtona replied:
in Zeile 26 wird das geprüft

MarcusRiemer commented on app/views/import_mappings/apply_mapping.html.haml:
```diff
@@ -0,0 +1,82 @@
+
+      - if @overwrite
+        %p.hint
+          Overwriting places with the same title is enabled.
```
Same key oder? Und die Spalten für den Key am Besten ausgeben.

birtona replied:
changed: https://github.com/a-thousand-channels/ORTE-backend/pull/406/commits/a09ce755184a10d05b52ae8c5554d215c4949bb8

MarcusRiemer commented on lib/imports/mapping_csv_importer.rb:
```diff
@@ -0,0 +1,99 @@
+  class MappingCsvImporter
+    attr_reader :valid_rows, :invalid_rows, :duplicate_rows, :errored_rows, :ambiguous_rows, :unprocessable_fields, :error
+
+    REQUIRED_FIELDS = %w[title lat lon].freeze
```
Technically this is for the model to decide. But not sure whether https://stackoverflow.com/questions/58612194/how-do-i-get-a-list-of-required-fields-in-a-model-in-rails is really nicer.

birtona replied:
I don't think that the other variant is nicer, and as these are only 3 attributes, I would prefer to leave it like this

MarcusRiemer commented on lib/imports/mapping_csv_importer.rb:
```diff
@@ -0,0 +1,99 @@
+
+    REQUIRED_FIELDS = %w[title lat lon].freeze
+
+    ALLOWED_FIELDS = %w[title subtitle teaser text link startdate startdate_date startdate_time enddate enddate_date enddate_time lat lon location address zip city country published featured sensitive sensitive_radius shy imagelink layer_id icon_id relations_tos relations_froms tag_list].freeze
```
This imho should be populated from the model

birtona commented on app/models/import_mapping.rb:
```diff
@@ -5,15 +5,37 @@ class ImportMapping < ApplicationRecord
 
+  @@parsers = {
+    'trim' => ->(value) { value.strip },
+    'sanitize' => ->(value) { ActionController::Base.helpers.sanitize(value) }, # was soll genau machen?
```
changed

ut commented:
I tried out today this branch locally, selected a csv file and proceeded. The following error occured. It seems that mapping is a string, while line 21 expects something different, a hash.

ActionView::Template::Error (undefined method `find' for an instance of String):

Causes:
NoMethodError (undefined method `find' for an instance of String)
   18:       - @headers&.each do |header|
   19:         %tr
   20:           %td= header
   21:           - mapping = @import_mapping.mapping.find { |m| m['csv_column_name'] == header }
   22:           - unless mapping
   23:             - mapping = { 'csv_column_name' => header, 'model_property' => nil, 'parsers' => [], 'key' => false }
   24:             - @import_mapping.mapping << mapping

ut commented:
> I tried out today this branch locally, selected a csv file and proceeded. The following error occured. It seems that mapping is a string

same error on another machine, the error occurs only if the CSV is valid (sic!) no matter which. I used CSV files from spec/support/files

MarcusRiemer commented on app/helpers/import_context_helper.rb:
```diff
@@ -0,0 +1,52 @@
+
+  def self.read_importing_rows(file_name)
+    Rails.cache.read("import_context/#{file_name}/importing_rows")&.[](:rows)
+  end
```
```
  def self.write_importing_rows(file_name, rows)
    prev = Rails.cache.read("import_context/#{file_name}") # TODO: Error handling

    Rails.cache.write("import_context/#{file_name}", prev.merge(importing_rows: rows), expires_in: 1.week)
  end

  def self.read_importing_rows(file_name)
    Rails.cache.read("import_context/#{file_name}")&.[](:importing_rows)
  end
```

MarcusRiemer commented on lib/imports/mapping_csv_importer.rb:
```diff
@@ -0,0 +1,109 @@
+            parsing_errors[model_property] = e.message
+          end
+        end
+        if @layer
+          processed_row[:layer_id] = @layer.id
+        elsif @fallback_layer
+          processed_row[:layer_id] = @fallback_layer.id if @fallback_layer
+        end
```
Dies überschreibt eine potenziell mit speziellen Parsern verarbeitetes Ergebnis für eine `layer_id` Spalte. Stand jetzt kann man effektiv für `layer_id` keine Parser generieren.

birtona commented:
> I tried out today this branch locally, selected a csv file and proceeded. The following error occured. It seems that mapping is a string, while line 21 expects something different, a hash.
> 
> ```
> ActionView::Template::Error (undefined method `find' for an instance of String):
> 
> Causes:
> NoMethodError (undefined method `find' for an instance of String)
>    18:       - @headers&.each do |header|
>    19:         %tr
>    20:           %td= header
>    21:           - mapping = @import_mapping.mapping.find { |m| m['csv_column_name'] == header }
>    22:           - unless mapping
>    23:             - mapping = { 'csv_column_name' => header, 'model_property' => nil, 'parsers' => [], 'key' => false }
>    24:             - @import_mapping.mapping << mapping
> ```

@ut  I applied the fix you suggested: https://github.com/a-thousand-channels/ORTE-backend/pull/406/commits/dbf5833c709027d4ac8a781a731f4a58334dc820

ut commented on app/views/import_mappings/new.html.haml:
```diff
@@ -0,0 +1,122 @@
+     });
+  // handle row duplication
+  document.querySelector('#mapping').addEventListener('click', function(event) {
+    if (event.target.classList.contains('duplicate-row')) {
```
I like this approach, cleaner mapping sounds great.

With my suggestion I focused on user experience. I had the curious but cautious user in mind, that clicks on the plus button just to see what happens and then wants to undo this hastly and might get nervous if this is not possible :)

So I think both things would be good.  But no sweat because of the latter, I would put this idea to the UI todo list :)

birtona replied:
@ut I changed it like that, but now I stumbled upon the side effect, that when a field is removed automatically, the mapping won't be detected as matching mapping for the dropdown to select from (even for the same file it has been created with), because it is missing the mapping for the removed field (this happens for example if nothing is mapped to the annotations column). That is also the underlying reason that the pipeline failed now, because even the newly created mapping won't be detected as matching. So I guess I will change that back, and probably your idea to do allow removal via frontend will be better - maybe also for not duplicated rows, if implemented well tested.

ut replied:
I dont really get the reason, but what counts, that it sounds too complicated to solve. So please revert :)

ut replied:
Ah you already did

ut commented:
Hey, some quite late testing feedbacks from my collegues:

1.  during testing the wish occured, to save an already saved mapping. In any testrun, it ended with several mappings until everything worked, only the last mapping shall be kept. we wonder if overwriting might be a quick and stressless solution?

2. the feature to split one field into two (from "location" with lat/lon value to "lat" and "lon") works super! but in the above mentioned effort to edit mapping, go to import preview and then go back to improve something, the setting with the additional field got lost.
 
3. I tried to reimport a file from d., into my map. The importer warned me that it will overwrite existing places. ID was key=true. so maybe, the ID thing is so strong, that I could overwrite places in a different map?

4. d. had the problem, that the allocation of layer slugs only worked after they renamed the column from "Layer" to "Document" (hence any other name then "Layer") and that workaround worked. I myself could not reproduce this, but I couldn't add entries with a colum of slugs of the respective layers. but the preview stated "Layer must exist". I can provide you me test file if needed.

Long story short, 1) overwriting an existing mapping would be great to avoid orphaned mappings! 2)+3) are kind of low key considering the time 4) would be nice if you could verify if an how the layer work. Thank you

ut commented:
> 4. d. had the problem, that the allocation of layer slugs only worked after they renamed the column from "Layer" to "layer_id"  and that workaround worked.

There was a misunderstanding. "layer_id" ist the column header that works! In a second try I could reproduce this error: A map import on different layers was only pissible when the layer column has the name "layer_id". Any other title won't work

birtona commented:
> Hey, some quite late testing feedbacks from my collegues:
> 
> 1. during testing the wish occured, to save an already saved mapping. In any testrun, it ended with several mappings until everything worked, only the last mapping shall be kept. we wonder if overwriting might be a quick and stressless solution?
> 2. the feature to split one field into two (from "location" with lat/lon value to "lat" and "lon") works super! but in the above mentioned effort to edit mapping, go to import preview and then go back to improve something, the setting with the additional field got lost.
> 3. I tried to reimport a file from d., into my map. The importer warned me that it will overwrite existing places. ID was key=true. so maybe, the ID thing is so strong, that I could overwrite places in a different map?
> 4. d. had the problem, that the allocation of layer slugs only worked after they renamed the column from "Layer" to "Document" (hence any other name then "Layer") and that workaround worked. I myself could not reproduce this, but I couldn't add entries with a colum of slugs of the respective layers. but the preview stated "Layer must exist". I can provide you me test file if needed.
> 
> Long story short, 1) overwriting an existing mapping would be great to avoid orphaned mappings! 2)+3) are kind of low key considering the time 4) would be nice if you could verify if an how the layer work. Thank you

Hi @ut,

thanks for the feedback. I fixed all issues except 2), which compared to the others imho is mostly a question of convenience for rare use cases. Additionally, I fixed some issues I encountered on the way (see commit message from https://github.com/a-thousand-channels/ORTE-backend/pull/406/commits/931bb36a5cf8bc148f18c6a83ef9ade8facabdab for details)

ut commented:
Merged finally :) Will be deployed in production the next days.
